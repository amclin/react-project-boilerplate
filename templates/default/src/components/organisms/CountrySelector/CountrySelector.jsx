import React, { useState } from 'react'
import { useQuery } from '@apollo/react-hooks'
import COUNTRIES_QUERY from '../../../graphql/queries/countries'
import Country from '../../atoms/Country'

const CountrySelector = () => {
  const { loading, error, data } = useQuery(COUNTRIES_QUERY)
  const [selectedCountryCode, handleSelectedCountryCode] = useState('')

  if (error) {
    return (
      <div data-testid="error-div" className="country-selector">
        Error loading countries!
      </div>
    )
  }
  if (loading) {
    return (
      <div data-testid="loading-div" className="country-selector">
        Loading Countries...
      </div>
    )
  }

  const { countries = [] } = data

  return (
    <div data-testid="selector-container" className="country-selector">
      <div>Choose a country!</div>
      <select
        value={selectedCountryCode}
        onChange={e => handleSelectedCountryCode(e.target.value)}
        data-testid="countries"
      >
        {countries.map(country => (
          <option
            key={country.code}
            value={country.code}
            data-testid="country-option"
          >
            {country.name}
          </option>
        ))}
      </select>
      <div data-testid="chosen-text">
        {selectedCountryCode ? 'Your chosen country is:' : 'No country chosen.'}
      </div>
      {selectedCountryCode && <Country code={selectedCountryCode} />}
    </div>
  )
}

export default CountrySelector
