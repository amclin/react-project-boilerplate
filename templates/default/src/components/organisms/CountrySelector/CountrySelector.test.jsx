import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { useQuery } from '@apollo/react-hooks'
import CountrySelector from '.'

jest.mock('@apollo/react-hooks')

describe('CountrySelector', () => {
  let mockQueryResponse = {}

  it('should render error output on error', () => {
    mockQueryResponse = {
      error: {
        graphQLErrors: []
      }
    }
    useQuery.mockReturnValue(mockQueryResponse)
    const { getByTestId } = render(<CountrySelector />)

    expect(getByTestId('error-div').textContent).toBe(
      'Error loading countries!'
    )
  })
  it('should render loading output when loading', () => {
    mockQueryResponse = {
      loading: true
    }
    useQuery.mockReturnValue(mockQueryResponse)
    const { getByTestId } = render(<CountrySelector />)

    expect(getByTestId('loading-div').textContent).toBe('Loading Countries...')
  })

  it('should render with broken data', () => {
    mockQueryResponse = {
      loading: false,
      data: {}
    }
    useQuery.mockReturnValue(mockQueryResponse)
    const { getByTestId } = render(<CountrySelector />)
    const container = getByTestId('selector-container')

    // eslint-disable-next-line
    expect(container).toExist
  })

  it('should render and let you choose a country', () => {
    mockQueryResponse = {
      loading: false,
      data: {
        countries: [
          {
            name: 'United States',
            code: 'US',
            emoji: 'E'
          },
          {
            name: 'France',
            code: 'FR',
            emoji: 'F'
          },
          {
            name: 'Canada',
            code: 'CA',
            emoji: 'C'
          }
        ]
      }
    }
    useQuery.mockReturnValue(mockQueryResponse)
    const {
      asFragment,
      getByTestId,
      getAllByTestId,
      getByDisplayValue
    } = render(<CountrySelector />)
    const countries = getByTestId('countries')
    const countryOptions = getAllByTestId('country-option')
    const chosenText = getByTestId('chosen-text')

    expect(countryOptions.length).toBe(3)
    expect(countries.value).toBe('US')
    expect(chosenText.textContent).toBe('No country chosen.')

    fireEvent.change(getByDisplayValue(/United States/i), {
      target: { value: 'FR' }
    })

    expect(countries.value).toBe('FR')
    expect(chosenText.textContent).toBe('Your chosen country is:')

    fireEvent.change(getByDisplayValue(/France/i), {
      target: { value: 'CA' }
    })

    expect(countries.value).toBe('CA')
    expect(chosenText.textContent).toBe('Your chosen country is:')

    expect(asFragment()).toMatchSnapshot()
  })
})
