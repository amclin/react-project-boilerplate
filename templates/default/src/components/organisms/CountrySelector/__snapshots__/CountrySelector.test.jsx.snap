// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountrySelector should render and let you choose a country 1`] = `
<DocumentFragment>
  <div
    class="country-selector"
    data-testid="selector-container"
  >
    <div>
      Choose a country!
    </div>
    <select
      data-testid="countries"
    >
      <option
        data-testid="country-option"
        value="US"
      >
        United States
      </option>
      <option
        data-testid="country-option"
        value="FR"
      >
        France
      </option>
      <option
        data-testid="country-option"
        value="CA"
      >
        Canada
      </option>
    </select>
    <div
      data-testid="chosen-text"
    >
      Your chosen country is:
    </div>
    <div
      class="country"
      data-testid="no-data-div"
    >
      No Country Data
    </div>
  </div>
</DocumentFragment>
`;
